src/framework/MOM_horizontal_regridding.F90

```diff
@@ -419,117 +425,144 @@ subroutine horiz_interp_and_extrap_tracer_record(filename, varnam,  conversion,
-
   roundoff = 3.0*EPSILON(missing_value)
 
   ! loop through each data level and interpolate to model grid.
   ! after interpolating, fill in points which will be needed
   ! to define the layers
   do k=1,kd
     write(laynum,'(I8)') k ; laynum = adjustl(laynum)
+    mask_in=0.0
+    if (is_ongrid) then
+       start(1) = is+G%HI%idg_offset ; start(2) = js+G%HI%jdg_offset ; start(3) = k
+       count(1) = ie-is+1 ; count(2) = je-js+1; count(3) = 1
+       rcode = NF90_GET_VAR(ncid,varid, tr_in, start, count)
+       if (rcode /= 0) call MOM_error(FATAL,"hinterp_and_extract_from_Fie: "//&
+            "error reading level "//trim(laynum)//" of variable "//&
+            trim(varnam)//" in file "// trim(filename))
+
+       do j=js,je
+         do i=is,ie
+           if (abs(tr_in(i,j)-missing_value) > abs(roundoff*missing_value)) then
+              mask_in(i,j) = 1.0
+              tr_in(i,j) = (tr_in(i,j)*scale_factor+add_offset) * conversion
+           else
+              tr_in(i,j) = missing_value
+           endif
+         enddo
+       enddo
 
-    if (is_root_pe()) then
-      start = 1 ; start(3) = k ; count(:) = 1 ; count(1) = id ; count(2) = jd
-      rcode = NF90_GET_VAR(ncid,varid, tr_in, start, count)
-      if (rcode /= 0) call MOM_error(FATAL,"hinterp_and_extract_from_Fie: "//&
-           "error reading level "//trim(laynum)//" of variable "//&
-           trim(varnam)//" in file "// trim(filename))
+    else
+       if (is_root_pe()) then
+          start = 1 ; start(3) = k ; count(:) = 1 ; count(1) = id ; count(2) = jd
+          rcode = NF90_GET_VAR(ncid,varid, tr_in, start, count)
+          if (rcode /= 0) call MOM_error(FATAL,"hinterp_and_extract_from_Fie: "//&
+               "error reading level "//trim(laynum)//" of variable "//&
+               trim(varnam)//" in file "// trim(filename))
+
+          if (add_np) then
+             last_row(:)=tr_in(:,jd); pole=0.0;npole=0.0
+             do i=1,id
+               if (abs(tr_in(i,jd)-missing_value) > abs(roundoff*missing_value)) then
+                  pole = pole+last_row(i)
+                  npole = npole+1.0
+               endif
+             enddo
+             if (npole > 0) then
+                pole=pole/npole
+             else
+                pole=missing_value
+             endif
+             tr_inp(:,1:jd) = tr_in(:,:)
+             tr_inp(:,jdp) = pole
+          else
+             tr_inp(:,:) = tr_in(:,:)
+          endif
+       endif
 
-      if (add_np) then
-         last_row(:)=tr_in(:,jd); pole=0.0;npole=0.0
+       call mpp_sync()
+       call mpp_broadcast(tr_inp, id*jdp, root_PE())
+       call mpp_sync_self()
+
+       do j=1,jdp
          do i=1,id
-            if (abs(tr_in(i,jd)-missing_value) > abs(roundoff*missing_value)) then
-               pole = pole+last_row(i)
-               npole = npole+1.0
-            endif
+           if (abs(tr_inp(i,j)-missing_value) > abs(roundoff*missing_value)) then
```

Comment:
Is there a way to read a mask from a netcdf file? We have `mask=` when writing and the reverse is what we should be doing here.

Reply:
If a mask is available, we could read it. Otherwise we need to generate it using missing/fill values as we're doing.  @raphaeldussin and I were discussing related issue regarding horizontal regridding/remapping and we agree that generating remapping weights offline is a good option which could be applied to both initialization and forcing data.